Allow head requests for prefixes without trailing "/"

nitisht opened the pull request:
## Description
Reverted some of the changes from #5900 to allow Spark to work with Minio 
server in FS mode. 

## Motivation and Context
To ensure Spark runs with Minio server in FS mode.

## How Has This Been Tested?
Manually with Spark and Minio FS server

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

harshavardhana reviewed (changes requested):
As suggested we should avoid using `fsIsDir` instead use `fs.isObjectDir` 

```diff
diff --git a/cmd/fs-v1-helpers.go b/cmd/fs-v1-helpers.go
index a3cd1f95..bb650b67 100644
--- a/cmd/fs-v1-helpers.go
+++ b/cmd/fs-v1-helpers.go
@@ -225,15 +225,6 @@ func fsStatDir(ctx context.Context, statDir string) (os.FileInfo, error) {
        return fi, nil
 }
 
-// Returns if the dirPath is a directory.
-func fsIsDir(ctx context.Context, dirPath string) bool {
-       fi, err := fsStat(ctx, dirPath)
-       if err != nil {
-               return false
-       }
-       return fi.IsDir()
-}
-
 // Lookup if file exists, returns file attributes upon success.
 func fsStatFile(ctx context.Context, statFile string) (os.FileInfo, error) {
        fi, err := fsStat(ctx, statFile)
diff --git a/cmd/fs-v1-helpers_test.go b/cmd/fs-v1-helpers_test.go
index 0df862a3..227e495b 100644
--- a/cmd/fs-v1-helpers_test.go
+++ b/cmd/fs-v1-helpers_test.go
@@ -548,18 +548,6 @@ func TestFSRemoveMeta(t *testing.T) {
        }
 }
 
-func TestFSIsDir(t *testing.T) {
-       dirPath, err := ioutil.TempDir(globalTestTmpDir, "minio-")
-       if err != nil {
-               t.Fatalf("Unable to create tmp directory %s", err)
-       }
-       defer os.RemoveAll(dirPath)
-
-       if !fsIsDir(context.Background(), dirPath) {
-               t.Fatalf("Expected %s to be a directory", dirPath)
-       }
-}
-
 func TestFSIsFile(t *testing.T) {
        dirPath, err := ioutil.TempDir(globalTestTmpDir, "minio-")
        if err != nil {
diff --git a/cmd/fs-v1.go b/cmd/fs-v1.go
index ececb7b2..eba4a171 100644
--- a/cmd/fs-v1.go
+++ b/cmd/fs-v1.go
@@ -563,9 +563,10 @@ func (fs *FSObjects) defaultFsJSON(object string) fsMetaV1 {
 // getObjectInfo - wrapper for reading object metadata and constructs ObjectInfo.
 func (fs *FSObjects) getObjectInfo(ctx context.Context, bucket, object string) (oi ObjectInfo, e error) {
        fsMeta := fsMetaV1{}
+       isDir := fs.isObjectDir(bucket, object)
        if hasSuffix(object, slashSeparator) {
                // Since we support PUT of a "directory" object, we allow HEAD.
-               if !fsIsDir(ctx, pathJoin(fs.fsPath, bucket, object)) {
+               if !isDir {
                        return oi, errFileNotFound
                }
                fi, err := fsStatDir(ctx, pathJoin(fs.fsPath, bucket, object))
@@ -573,6 +574,8 @@ func (fs *FSObjects) getObjectInfo(ctx context.Context, bucket, object string) (
                        return oi, err
                }
                return fsMeta.ToObjectInfo(bucket, object, fi), nil
+       } else if isDir {
+               return oi, errFileNotFound
        }
 
        fsMetaPath := pathJoin(fs.fsPath, minioMetaBucket, bucketMetaPrefix, bucket, object, fs.metaJSONFile)
@@ -623,11 +626,7 @@ func (fs *FSObjects) getObjectInfoWithLock(ctx context.Context, bucket, object s
        }
 
        if _, err := fs.statBucketDir(ctx, bucket); err != nil {
-               return oi, toObjectErr(err, bucket)
-       }
-
-       if strings.HasSuffix(object, slashSeparator) && !fs.isObjectDir(bucket, object) {
-               return oi, errFileNotFound
+               return oi, err
        }
 
        return fs.getObjectInfo(ctx, bucket, object)
```

nitisht committed:
Throw 404 for head requests for prefixes without trailing "/"

Minio server returns 403 (access denied) for head requests to prefixes
without trailing "/", this is different from S3 behaviour. S3 returns
404 in such cases.

Fixes #6080

vadmeste reviewed (approved):
LGTM
